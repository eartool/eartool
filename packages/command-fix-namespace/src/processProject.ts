import {
  addSingleFileReplacementsForRenames,
  processReplacements,
  ProjectContext,
  ReplacementsWrapperForContext,
  SimpleReplacements,
} from "@eartool/replacements";
import type { PackageExportRename } from "@eartool/replacements";
import { dropDtsFiles, organizeImportsOnFiles } from "@eartool/utils";
import type { PackageContext, PackageName } from "@eartool/utils";
import type { Logger } from "pino";
import { calculateNamespaceLikeRemovals } from "./calculateNamespaceLikeRemovals.js";
import { calculateNamespaceRemovals } from "./calculateNamespaceRemovals.js";

export interface Status {
  totalWorkUnits: number;
  completedWorkUnits: number;
  stage: "analyzing" | "writing" | "organizing";
}

export interface BaseOptions {
  dryRun: boolean;
  removeNamespaces: boolean;
  removeFauxNamespaces: boolean;
  organizeImports: boolean;
}

export interface ProcessProjectOpts extends BaseOptions {
  logger: Logger;
  updateState?: (data: Status) => void;
  additionalRenames?: Map<PackageName, PackageExportRename[]>;
}

/**
 * Process a project, removing namespaces and organizing imports.
 * @param project  The project to process
 * @param param1  Options
 * @returns
 */
export async function processProject(
  packageContext: PackageContext,
  {
    dryRun = false,
    logger,
    updateState = (_data) => undefined,
    additionalRenames,
    removeNamespaces,
    removeFauxNamespaces,
    organizeImports: shouldOrganizeImports,
  }: ProcessProjectOpts,
) {
  const { project } = packageContext;

  dropDtsFiles(project);
  const context = new ProjectContext(
    packageContext.project,
    logger,
    packageContext.packagePath,
    packageContext.packageName,
    packageContext.packageJson,
  );
  const totalFiles = project.getSourceFiles().length;
  // Three stages:
  // * analyzing:
  //   * if `removeNamespaces`: add `totalFiles`
  //   * if `additionalRenames`: add `totalFiles`
  // * organizing `[...changedFiles].length` work.
  // * writing  `[]
  function calculateTotalWorkUnits(changedFilesCount: number) {
    return (
      (removeNamespaces ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (dryRun ? 0 : changedFilesCount) +
      (shouldOrganizeImports ? changedFilesCount : 0)
    );
  }

  // Assume all files change for now
  let totalWorkUnits = calculateTotalWorkUnits(totalFiles);

  let completedWorkUnits = 0;
  // TODO: Rename totalFiles here to totalWorkUnits or similar
  updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });

  logger.debug("Running with opts %o", {
    dryRun,
    removeNamespaces,
    removeFauxNamespaces,
    organizeImports: shouldOrganizeImports,
    additionalRenames: [...(additionalRenames?.entries() ?? [])],
  });

  for (const sf of project.getSourceFiles()) {
    const replacements = new ReplacementsWrapperForContext(context);

    if (removeNamespaces) {
      calculateNamespaceRemovals(sf, context, replacements);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (removeFauxNamespaces) {
      calculateNamespaceLikeRemovals(sf, context, replacements);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (additionalRenames) {
      const replacements = new SimpleReplacements(logger);
      addSingleFileReplacementsForRenames(context, sf, additionalRenames, replacements, dryRun);
      for (const r of replacements.getReplacementsArray()) {
        context.addReplacement(r);
      }
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }
  }

  // actually updates files in project!
  const changedFiles = [...processReplacements(project, context.getReplacements())];
  totalWorkUnits = calculateTotalWorkUnits(changedFiles.length);

  if (shouldOrganizeImports) {
    logger.debug("Organizing imports");
    updateState({ totalWorkUnits, completedWorkUnits, stage: "organizing" });
    organizeImportsOnFiles(project, changedFiles);
  }

  completedWorkUnits += changedFiles.length; // TODO make this granular?
  updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });

  if (dryRun) {
    logger.info("DRY RUN");
  } else {
    logger.info("Saving");
    await project.save();
    completedWorkUnits += changedFiles.length; // TODO make this granular?
    updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });
  }

  logger.debug("Exported renames %o", context.getRecordedRenames());

  return {
    exportedRenames: context.getRecordedRenames(),
  };
}
