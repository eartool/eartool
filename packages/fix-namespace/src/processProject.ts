import type { Node, SourceFile } from "ts-morph";
import type { Project } from "ts-morph";
import { calculateNamespaceRemovals } from "./calculateNamespaceRemovals.js";
import type { Logger } from "pino";
import { ProjectContext } from "./Context.js";
import { processReplacements } from "./replacements/processReplacements.js";
import { dropDtsFiles } from "./utils/tsmorph/dropDtsFiles.js";
import { organizeImportsOnFiles } from "./utils/tsmorph/organizeImportsOnFiles.js";
import type { PackageExportRename } from "./replacements/PackageExportRename.js";
import type { PackageName } from "./PackageName.js";
import { addSingleFileReplacementsForRenames } from "./replacements/addSingleFileReplacementsForRenames.js";
import type { Replacement } from "./replacements/Replacement.js";
import { ReplacementsWrapper } from "./ReplacementsWrapper.js";
import { calculateNamespaceLikeRemovals } from "./calculateNamespaceLikeRemovals.js";

export interface Status {
  totalWorkUnits: number;
  completedWorkUnits: number;
  stage: "analyzing" | "writing" | "organizing";
}

export interface BaseOptions {
  dryRun: boolean;
  removeNamespaces: boolean;
  removeFauxNamespaces: boolean;
  organizeImports: boolean;
}

export interface ProcessProjectOpts extends BaseOptions {
  logger: Logger;
  updateState?: (data: Status) => void;
  additionalRenames?: Map<PackageName, PackageExportRename[]>;
}

/**
 * Process a project, removing namespaces and organizing imports.
 * @param project  The project to process
 * @param param1  Options
 * @returns
 */
export async function processProject(
  project: Project,
  {
    dryRun = false,
    logger,
    updateState = (_data) => undefined,
    additionalRenames,
    removeNamespaces,
    removeFauxNamespaces,
    organizeImports: shouldOrganizeImports,
  }: ProcessProjectOpts
) {
  dropDtsFiles(project);
  const context = new ProjectContext(project, logger);
  const totalFiles = project.getSourceFiles().length;
  // Three stages:
  // * analyzing:
  //   * if `removeNamespaces`: add `totalFiles`
  //   * if `additionalRenames`: add `totalFiles`
  // * organizing `[...changedFiles].length` work.
  // * writing  `[]
  function calculateTotalWorkUnits(changedFilesCount: number) {
    return (
      (removeNamespaces ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (dryRun ? 0 : changedFilesCount) +
      (shouldOrganizeImports ? changedFilesCount : 0)
    );
  }

  // Assume all files change for now
  let totalWorkUnits = calculateTotalWorkUnits(totalFiles);

  let completedWorkUnits = 0;
  // TODO: Rename totalFiles here to totalWorkUnits or similar
  updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });

  logger.debug("Running with opts %o", {
    dryRun,
    removeNamespaces,
    additionalRenames: [...(additionalRenames?.entries() ?? [])],
  });

  for (const sf of project.getSourceFiles()) {
    if (removeNamespaces) {
      calculateNamespaceRemovals(sf, context);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (removeFauxNamespaces) {
      const replacements = new ReplacementsWrapper(context);
      calculateNamespaceLikeRemovals(sf, replacements);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (additionalRenames) {
      const replacements: Replacement[] = [];
      addSingleFileReplacementsForRenames(sf, additionalRenames, replacements, logger);
      for (const r of replacements) {
        context.addReplacement(r);
      }
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }
  }

  // actually updates files in project!
  const changedFiles = [...processReplacements(project, context.getReplacements())];
  totalWorkUnits = calculateTotalWorkUnits(changedFiles.length);

  if (shouldOrganizeImports) {
    logger.debug("Organizing imports");
    updateState({ totalWorkUnits, completedWorkUnits, stage: "organizing" });
    organizeImportsOnFiles(project, changedFiles);
  }

  completedWorkUnits += changedFiles.length; // TODO make this granular?
  updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });

  if (dryRun) {
    logger.info("DRY RUN");
  } else {
    logger.info("Saving");
    await project.save();
    completedWorkUnits += changedFiles.length; // TODO make this granular?
    updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });
  }

  return {
    exportedRenames: context.getRecordedRenames(),
  };
}

export function getFilePath(filePath: string | Node) {
  return typeof filePath == "string" ? filePath : filePath.getSourceFile().getFilePath();
}
