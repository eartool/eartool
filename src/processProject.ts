import { Node, SyntaxKind, type Project, type SourceFile, NamedImports } from "ts-morph";
import { calculateNamespaceRemovals } from "./calculateNamespaceRemovals.js";
import type { Logger } from "pino";
import { ProjectContext } from "./Context.js";
import { processReplacements } from "./replacements/processReplacements.js";
import { dropDtsFiles } from "./utils/tsmorph/dropDtsFiles.js";
import { organizeImportsOnFiles } from "./utils/tsmorph/organizeImportsOnFiles.js";
import type { PackageExportRename } from "./replacements/PackageExportRename.js";
import type { PackageName } from "./PackageName.js";
import { addSingleFileReplacementsForRenames } from "./replacements/addSingleFileReplacementsForRenames.js";
import type { Replacement } from "./replacements/Replacement.js";
import * as Assert from "assert";
import { ReplacementsWrapper } from "./ReplacementsWrapper.js";
import type { Replacements } from "./replacements/Replacements.js";
import { isNamespaceLike } from "./utils/tsmorph/isNamespaceLike.js";
import { isAnyOf } from "@reduxjs/toolkit";

export interface Status {
  totalWorkUnits: number;
  completedWorkUnits: number;
  stage: "analyzing" | "writing" | "organizing";
}

export interface ProcessProjectOpts {
  dryRun?: boolean;
  logger: Logger;
  removeNamespaces: boolean;
  updateState?: (data: Status) => void;
  additionalRenames?: Map<PackageName, PackageExportRename[]>;
}

export async function processProject(
  project: Project,
  {
    dryRun = false,
    logger,
    updateState = (_data) => undefined,
    additionalRenames,
    removeNamespaces,
  }: ProcessProjectOpts
) {
  dropDtsFiles(project);
  const context = new ProjectContext(project, logger);
  const totalFiles = project.getSourceFiles().length;
  // Three stages:
  // * analyzing:
  //   * if `removeNamespaces`: add `totalFiles`
  //   * if `additionalRenames`: add `totalFiles`
  // * organizing `[...changedFiles].length` work.
  // * writing  `[]
  function calculateTotalWorkUnits(changedFilesCount: number) {
    return (
      (removeNamespaces ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (dryRun ? 0 : changedFilesCount) +
      changedFilesCount
    );
  }

  // Assume all files change for now
  let totalWorkUnits = calculateTotalWorkUnits(totalFiles);

  let completedWorkUnits = 0;
  // TODO: Rename totalFiles here to totalWorkUnits or similar
  updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });

  logger.debug("Running with opts %o", {
    dryRun,
    removeNamespaces,
    additionalRenames: [...(additionalRenames?.entries() ?? [])],
  });

  for (const sf of project.getSourceFiles()) {
    if (removeNamespaces) {
      calculateNamespaceRemovals(sf, context);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    const replacements = new ReplacementsWrapper(context);
    calculateNamespaceLikeRemovals(sf, replacements);

    if (additionalRenames) {
      const replacements: Replacement[] = [];
      addSingleFileReplacementsForRenames(sf, additionalRenames, replacements, logger);
      for (const r of replacements) {
        context.addReplacement(r);
      }
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }
  }

  // actually updates files in project!
  const changedFiles = [...processReplacements(project, context.getReplacements())];
  totalWorkUnits = calculateTotalWorkUnits(changedFiles.length);

  logger.debug("Organizing imports");
  updateState({ totalWorkUnits, completedWorkUnits, stage: "organizing" });
  organizeImportsOnFiles(project, changedFiles);

  completedWorkUnits += changedFiles.length; // TODO make this granular?
  updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });

  if (dryRun) {
    logger.info("DRY RUN");
  } else {
    logger.info("Saving");
    await project.save();
    completedWorkUnits += changedFiles.length; // TODO make this granular?
    updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });
  }

  return {
    exportedRenames: context.getRecordedRenames(),
  };
}

export function getFilePath(filePath: string | Node) {
  return typeof filePath == "string" ? filePath : filePath.getSourceFile().getFilePath();
}

function calculateNamespaceLikeRemovals(sf: SourceFile, replacements: Replacements) {
  // TODO: Only perform task if its the only export
  // TODO: Should we check the filename too?
  // TODO: Check for collisions?

  for (const statement of sf.getStatements()) {
    if (!isNamespaceLike(statement)) continue;
    const varDecl = statement.getDeclarations()[0];

    const visitedSpecifiers = new Set();
    for (const refIdentifier of varDecl.findReferencesAsNodes()) {
      // alias import nodes show up twice for some reason
      // so we need to account for that

      const specifier = refIdentifier.getParentIf(
        isAnyOf(Node.isExportSpecifier, Node.isImportSpecifier)
      );
      if (!specifier) continue;
      if (visitedSpecifiers.has(specifier)) continue;

      const named = specifier.getParentIfOrThrow(isAnyOf(Node.isNamedExports, Node.isNamedImports));
      Assert.ok(named.getElements().length == 1);
      const varName = (specifier.getAliasNode() ?? specifier.getNameNode()).getText();
      replacements.replaceNode(named, `* as ${varName}`);
      visitedSpecifiers.add(specifier);
    }

    const syntaxList = varDecl.getInitializer().getExpression().getChildSyntaxList()!;

    // Drop `export const Name = {`
    replacements.remove(sf, statement.getStart(), syntaxList.getFullStart());

    // drop `} as const;`
    const closeBrace = syntaxList.getNextSiblingIfKindOrThrow(SyntaxKind.CloseBraceToken);
    replacements.remove(sf, closeBrace.getStart(), statement.getEnd());

    for (const q of varDecl.getInitializer().getExpression().getProperties()) {
      if (Node.isMethodDeclaration(q)) {
        replacements.insertBefore(q, "export function ");
      } else {
        replacements.addReplacement(
          sf,
          q.getStart(),
          q.getFirstChildByKindOrThrow(SyntaxKind.ColonToken).getEnd(),
          `export const ${q.getName()} = `
        );
      }
      replacements.removeNextSiblingIfComma(q);
    }
  }
}
