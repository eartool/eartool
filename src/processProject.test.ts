import { describe, expect, it } from "@jest/globals";
import { format } from "prettier";
import { Node, Project, SourceFile } from "ts-morph";
import { processProject } from "./processProject.js";
import { pino } from "pino";
import pinoPretty from "pino-pretty";

function formatTestTypescript(src: string) {
  return format(src, { parser: "typescript", tabWidth: 2, useTabs: false });
}

const cases = [
  {
    name: "redeclare export",
    inputs: new Map([
      [
        "foo.tsx",
        `
        export namespace Foo {
          export function bar() {
            return 5;
          }

          export function baz() {
            return 5;
          }

          export type Thing = string;
        }
        `,
      ],
      [
        "index.ts",
        `
      export {Foo} from "./foo";
      `,
      ],
    ]),
    outputs: new Map([
      [
        "foo.tsx",
        `
          export function barOfFoo() {
           return 5;
          }

          export function bazOfFoo() {
            return 5;
           }
           
           export type ThingOfFoo = string;
      `,
      ],
      [
        "index.ts",
        `
    export {barOfFoo, bazOfFoo, type ThingOfFoo} from "./foo";
    `,
      ],
    ]),
  },
  {
    name: "function invoke within namespace",
    inputs: new Map([
      [
        "foo.tsx",
        `
        export namespace Foo {
          export function bar() {
            baz();
          }

          export function baz() {
            return 5;
          }
        }
        `,
      ],
    ]),
  },
  {
    name: "combined types",
    inputs: new Map([
      [
        "foo.tsx",
        `export namespace AssociatedMapSection {
        export interface OgreProps {
          properties: Property<any>[];
        }
      
        export interface ReduxProps {
          appRealmId: RealmId;
          mapConfig: GaiaMapConfig;
        }
      
        export interface State {
          map?: MapSearchResult;
          user?: IAcmeUser;
          isLoading: boolean;
        }
      
        export type Props = OverviewObjectMinProps & OgreProps & ReduxProps;
      }`,
      ],
    ]),
  },
  {
    name: "works nicely with interfaces",
    // We cant break foo out in this case cause im too lazy to implement this another way.
    inputs: new Map([
      [
        "foo.tsx",
        `
            export namespace Foo {
                export interface Props { what: number }
            }

            class Foo extends React.Component<Foo.Props> {

            }
            `,
      ],
    ]),
  },
  {
    name: "dont explode if error",
    // We cant break foo out in this case cause im too lazy to implement this another way.
    inputs: new Map([
      [
        "foo.ts",
        `
            const foo = 5;
            
            export namespace Wat {
                export const foo = 6;
            }
            `,
      ],
    ]),
  },
  {
    name: "simple",
    inputs: new Map([
      [
        "foo.ts",
        `
        const foo = 5;
        
        export namespace Wat {
            export const aasdf = 3;
            export const second = 5;

            export const thirdSpaced = 56;

            // Foo
            export const fourthWithComment = 555;
        }
        `,
      ],
    ]),
  },
  {
    name: "rename in other file",
    inputs: new Map([
      [
        "wat.ts",
        `
        
        export namespace Wat {
            export const key = 3;
            export function f() { return 5; }

            export class Foo {}

            export type Baz = string;
        }
        `,
      ],
      [
        "refWat.ts",
        `import {Wat} from "./wat";
        console.log(Wat.key);
        console.log(Wat.f());
        console.log(new Wat.Foo());
        const f: Wat.Baz = "hi";
        `,
      ],
    ]),
  },
];

describe("processProject", () => {
  it.each(cases)("$name", async ({ inputs, outputs }) => {
    const project = new Project({
      useInMemoryFileSystem: true,
      skipAddingFilesFromTsConfig: true,
    });
    for (const [name, contents] of inputs) {
      project.createSourceFile(name, contents);
    }

    const logger = pino(
      {
        level: "trace",
        serializers: {
          ...pino.stdSerializers,
          foo: (n: Node) => `${n.getSourceFile().getFilePath()}:${n.getStartLineNumber()}`,
        },
        hooks: {
          logMethod: function logMethod([msg, ...args], method, foo) {
            args = args.map((a: any) => maybeConvertNodeToFileAndLineNum(a));
            // console.log([msg, ...args]);
            method.apply(this, [msg, ...args]);
          },
        },
      },
      pinoPretty.default({
        colorize: true,
        sync: true,
      })
    );

    await processProject(project, {
      logger,
    });

    const fs = project.getFileSystem();

    const output = project
      .getSourceFiles()
      .map((sf) => {
        const filePath = sf.getFilePath();
        return formatTestTypescript(
          `
        //

        //
        // PATH: '${filePath}'
        //
        ${fs.readFileSync(filePath)}
        `
        );
      })
      .join();

    expect(output).toMatchSnapshot();
  });
});
function maybeConvertNodeToFileAndLineNum(a: any): any {
  if (a instanceof Node) {
    return `${a.getSourceFile().getFilePath()}:${a.getStartLineNumber()}`;
  }

  return a;
}
