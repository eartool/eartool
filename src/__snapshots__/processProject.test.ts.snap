// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processProject combined types 1`] = `
"//

//
// PATH: '/foo.tsx'
//
export interface AssociatedMapSectionOgreProps {
  properties: Property<any>[];
}

export interface AssociatedMapSectionReduxProps {
  appRealmId: RealmId;
  mapConfig: GaiaMapConfig;
}

export interface AssociatedMapSectionState {
  map?: MapSearchResult;
  user?: IAcmeUser;
  isLoading: boolean;
}

export type AssociatedMapSectionProps = OverviewObjectMinProps &
  AssociatedMapSectionOgreProps &
  AssociatedMapSectionReduxProps;
"
`;

exports[`processProject dont explode if error 1`] = `
"//

//
// PATH: '/foo.ts'
//

const foo = 5;

export namespace Wat {
  export const foo = 6;
}
"
`;

exports[`processProject function invoke within namespace 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export function barOfFoo() {
  bazOfFoo();
}

export function bazOfFoo() {
  return 5;
}
"
`;

exports[`processProject redeclare export 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export function barOfFoo() {
  return 5;
}

export function bazOfFoo() {
  return 5;
}

export type ThingOfFoo = string;
,//

//
// PATH: '/index.ts'
//

export { barOfFoo, bazOfFoo, type ThingOfFoo } from "./foo";
"
`;

exports[`processProject rename in other file 1`] = `
"//

//
// PATH: '/refWat.ts'
//
import { BazOfWat, FooOfWat, fOfWat, keyOfWat } from "./wat";

console.log(keyOfWat);
console.log(fOfWat());
console.log(new FooOfWat());
const f: BazOfWat = "hi";
,//

//
// PATH: '/wat.ts'
//

export const keyOfWat = 3;
export function fOfWat() {
  return 5;
}

export class FooOfWat {}

export type BazOfWat = string;
"
`;

exports[`processProject simple 1`] = `
"//

//
// PATH: '/foo.ts'
//

const foo = 5;

export const aasdfOfWat = 3;
export const secondOfWat = 5;

export const thirdSpacedOfWat = 56;

// Foo
export const fourthWithCommentOfWat = 555;
"
`;

exports[`processProject works nicely with interfaces 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface FooProps {
  what: number;
}

class Foo extends React.Component<FooProps> {}
"
`;
